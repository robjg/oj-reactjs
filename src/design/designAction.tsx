
import React from 'react';
import ReactDOM from 'react-dom';
import { ActionFactory, ActionContext, Action, contextSearch } from '../menu/actions';
import { ConfigurationOwner, ConfigPoint } from '../remote/ojremotes';
import { AddJobForm as AddJobForm } from './addjob';
import { ParserDesignModel } from './design';
import { DesignForm } from './designForm';
import { definedOrError } from '../main/util';
import { JAVA_OBJECT } from '../remote/java';

export class DesignActionFactory implements ActionFactory {

    createAction(actionContext: ActionContext): Action | null {

        const configOwner: ConfigurationOwner | null = contextSearch(actionContext, ConfigurationOwner);

        if (configOwner == null) {
            return null;
        }

        return {

            name: "Design",

            isEnabled: true,

            perform: () => {

                configOwner.formFor(actionContext.proxy)
                    .then(formText => {
                        function hide() {
                            formDiv.style.display = 'none';
                            ReactDOM.unmountComponentAtNode(formDiv);
                        }

                        const designModel = new ParserDesignModel(
                            {
                                formConfiguration: JSON.parse(formText),
                                saveAction: (config: any) => {
                                    hide();
                                    configOwner.replaceJson(actionContext.proxy,
                                        JSON.stringify(config));
                                },
                                newForm: (element: string, isComponent: boolean, propertyClass: string) =>
                                    configOwner.blankForm(isComponent, element, propertyClass)
                            });

                        const formDiv = definedOrError(document.getElementById('ojForm'), "No form div.");

                        ReactDOM.render(
                            <DesignForm designModel={designModel}
                                hideForm={hide} />,
                            formDiv);

                        formDiv.style.display = 'block';
                    });
            }

        }
    }
}

export class AddJobActionFactory implements ActionFactory {

    createAction(actionContext: ActionContext): Action | null {

        const configOwner: ConfigurationOwner | null = contextSearch(actionContext, ConfigurationOwner);

        if (configOwner == null) {
            return null;
        }
        const proxy = actionContext.proxy;

        if (!proxy.isA(ConfigPoint)) {
            return null;
        }

        const dragPoint: ConfigPoint = proxy.as(ConfigPoint);

        function launchDesignForm(tag: string): void {

            configOwner?.blankForm(true, tag, JAVA_OBJECT.name)
                .then(formText => {

                    function hide() {
                        formDiv.style.display = 'none';
                        ReactDOM.unmountComponentAtNode(formDiv);
                    }

                    const designModel = new ParserDesignModel(
                        {
                            formConfiguration: JSON.parse(formText),
                            saveAction: (config: any) => {
                                hide();
                                dragPoint?.paste(-1, JSON.stringify(config));
                            },
                            newForm: (element: string, isComponent: boolean, propertyClass: string) =>
                                configOwner.blankForm(isComponent, element, propertyClass)
                        });

                    const formDiv = definedOrError(document.getElementById('ojForm'), "No form div.");

                    ReactDOM.render(
                        <DesignForm designModel={designModel}
                            hideForm={hide} />,
                        formDiv);

                    formDiv.style.display = 'block';
                });
        }

        class Impl implements Action {

            readonly name = "Add Job";

            get isEnabled(): boolean {
                return dragPoint.isPasteSupported;
            }

            perform(): void {

                dragPoint.possibleChildren()
                    .then(possibleChildren => {
                        function hide() {
                            formDiv.style.display = 'none';
                            ReactDOM.unmountComponentAtNode(formDiv);
                        }

                        const formDiv = definedOrError(document.getElementById('ojForm'), "No form div.");

                        ReactDOM.render(
                            <AddJobForm options={possibleChildren}
                                launchDesignForm={launchDesignForm}
                                hideForm={hide} />,
                            formDiv);

                        formDiv.style.display = 'block';
                    });
            }

        }

        return new Impl();
    }
}

