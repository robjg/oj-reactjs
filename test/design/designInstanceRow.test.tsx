/**
 * @jest-environment jsdom
 */

import React, { ReactNode, useState } from 'react'
import { render, fireEvent, within, getByText } from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'

import { KeyedInstanceRow, InstanceRow } from '../../src/design/designForm'
import { DesignInstance, TextField } from '../../src/design/design';

function createFactory(): Map<string, DesignInstance> {

    const factory: Map<string, DesignInstance> = new Map();

    const apple = new DesignInstance('apple');
    apple.addItem(new TextField("colour"));

    factory.set('apple', apple);

    const orange = new DesignInstance('orange');
    orange.addItem(new TextField("seedless"));

    factory.set('orange', orange);

    return factory;
}

function SomeProperty(props: any) {

    const factory: Map<string, DesignInstance> = createFactory();

    const [instance, setInstance] = useState(factory.get('apple'));

    const changeTheInstance = (selected: string) => {
        setInstance(factory.get(selected));
    }

    const factoryKeys: string[] = Array.from(factory.keys());

    return <div>
        <p>The instance is {instance?.element}</p>
                <InstanceRow
                    instance={instance}
                    options={factoryKeys}
                    changeTheInstance={changeTheInstance}
                    idPrefix="foo" />
    </div>
}

test('Test Render Some Property', () => {

    const result = render(<SomeProperty />);

    expect(result).toMatchSnapshot();

    const expand = result.getByTestId("foo_expand");

    fireEvent.click(expand);

    expect(result).toMatchSnapshot();

     expect(result.getByTestId("foo_detail_colour")).toBeInTheDocument();

    const select = result.getByTestId("foo_select");

    fireEvent.change(select, { target: { value: "orange" } });

    expect(result).toMatchSnapshot();

    expect(result.getByText("The instance is orange")).toBeInTheDocument();
})

function SomeKeyedProperty(props: any) {

    const factory: Map<string, DesignInstance> = createFactory();

    const [instance, setInstance] = useState(factory.get('apple'));
    const [theKey, setTheKey] = useState('fruit');

    const changeTheKey = (theKey: string) => {
        setTheKey(theKey)
    }

    const changeTheInstance = (selected: string) => {
        setInstance(factory.get(selected));
    }

    const factoryKeys: string[] = Array.from(factory.keys());

    return <div>
        <p>Key is {theKey} and the instance is {instance?.element}</p>
                <KeyedInstanceRow
                    theKey={theKey}
                    instance={instance}
                    options={factoryKeys}
                    changeTheKey={changeTheKey}
                    changeTheInstance={changeTheInstance}
                    idPrefix="foo" />
    </div>
}

test('Test Render Some Key Property', () => {

    const result = render(<SomeKeyedProperty />);

    expect(result).toMatchSnapshot();

    const expand = result.getByTestId("foo_expand");

    fireEvent.click(expand);

    expect(result).toMatchSnapshot();

    expect(result.getByTestId("foo_detail_colour")).toBeInTheDocument();

    const select = result.getByTestId("foo_select");

    fireEvent.change(select, { target: { value: "orange" } });

    const key = result.getByTestId("foo_key");

    fireEvent.change(key, { target: { value: "snack" } })

    expect(result).toMatchSnapshot();

    expect(result.getByText("Key is snack and the instance is orange")).toBeInTheDocument();
})

