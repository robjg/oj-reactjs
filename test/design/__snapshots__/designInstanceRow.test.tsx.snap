// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Render Some Key Property 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <p>
          Key is 
          fruit
           and the instance is 
          apple
        </p>
        <div
          class="oj-design-row"
        >
          <span
            class="oj-design-key"
          >
            <input
              data-testid="foo_key"
              type="text"
              value="fruit"
            />
          </span>
          <span
            class="oj-design-select"
          >
            <select
              data-testid="foo_select"
            >
              <option
                value=""
              />
              <option
                value="apple"
              >
                apple
              </option>
              <option
                value="orange"
              >
                orange
              </option>
            </select>
          </span>
          <span
            class="oj-design-expand"
          >
            <button
              data-testid="foo_expand"
              type="button"
            >
              +
            </button>
          </span>
        </div>
        <div
          class="oj-design-detail"
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <p>
        Key is 
        fruit
         and the instance is 
        apple
      </p>
      <div
        class="oj-design-row"
      >
        <span
          class="oj-design-key"
        >
          <input
            data-testid="foo_key"
            type="text"
            value="fruit"
          />
        </span>
        <span
          class="oj-design-select"
        >
          <select
            data-testid="foo_select"
          >
            <option
              value=""
            />
            <option
              value="apple"
            >
              apple
            </option>
            <option
              value="orange"
            >
              orange
            </option>
          </select>
        </span>
        <span
          class="oj-design-expand"
        >
          <button
            data-testid="foo_expand"
            type="button"
          >
            +
          </button>
        </span>
      </div>
      <div
        class="oj-design-detail"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Test Render Some Key Property 2`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <p>
          Key is 
          fruit
           and the instance is 
          apple
        </p>
        <div
          class="oj-design-row"
        >
          <span
            class="oj-design-key"
          >
            <input
              data-testid="foo_key"
              type="text"
              value="fruit"
            />
          </span>
          <span
            class="oj-design-select"
          >
            <select
              data-testid="foo_select"
            >
              <option
                value=""
              />
              <option
                value="apple"
              >
                apple
              </option>
              <option
                value="orange"
              >
                orange
              </option>
            </select>
          </span>
          <span
            class="oj-design-expand"
          >
            <button
              data-testid="foo_expand"
              type="button"
            >
              -
            </button>
          </span>
        </div>
        <div
          class="oj-design-detail"
        >
          <div
            class="oj-design-internal-form"
          >
            <ul
              class="oj-form-group"
            >
              <li>
                <label>
                  colour
                  <input
                    data-testid="foo_detail_colour"
                    type="text"
                    value=""
                  />
                </label>
              </li>
            </ul>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <p>
        Key is 
        fruit
         and the instance is 
        apple
      </p>
      <div
        class="oj-design-row"
      >
        <span
          class="oj-design-key"
        >
          <input
            data-testid="foo_key"
            type="text"
            value="fruit"
          />
        </span>
        <span
          class="oj-design-select"
        >
          <select
            data-testid="foo_select"
          >
            <option
              value=""
            />
            <option
              value="apple"
            >
              apple
            </option>
            <option
              value="orange"
            >
              orange
            </option>
          </select>
        </span>
        <span
          class="oj-design-expand"
        >
          <button
            data-testid="foo_expand"
            type="button"
          >
            -
          </button>
        </span>
      </div>
      <div
        class="oj-design-detail"
      >
        <div
          class="oj-design-internal-form"
        >
          <ul
            class="oj-form-group"
          >
            <li>
              <label>
                colour
                <input
                  data-testid="foo_detail_colour"
                  type="text"
                  value=""
                />
              </label>
            </li>
          </ul>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Test Render Some Key Property 3`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <p>
          Key is 
          snack
           and the instance is 
          orange
        </p>
        <div
          class="oj-design-row"
        >
          <span
            class="oj-design-key"
          >
            <input
              data-testid="foo_key"
              type="text"
              value="snack"
            />
          </span>
          <span
            class="oj-design-select"
          >
            <select
              data-testid="foo_select"
            >
              <option
                value=""
              />
              <option
                value="apple"
              >
                apple
              </option>
              <option
                value="orange"
              >
                orange
              </option>
            </select>
          </span>
          <span
            class="oj-design-expand"
          >
            <button
              data-testid="foo_expand"
              type="button"
            >
              -
            </button>
          </span>
        </div>
        <div
          class="oj-design-detail"
        >
          <div
            class="oj-design-internal-form"
          >
            <ul
              class="oj-form-group"
            >
              <li>
                <label>
                  seedless
                  <input
                    data-testid="foo_detail_seedless"
                    type="text"
                    value=""
                  />
                </label>
              </li>
            </ul>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <p>
        Key is 
        snack
         and the instance is 
        orange
      </p>
      <div
        class="oj-design-row"
      >
        <span
          class="oj-design-key"
        >
          <input
            data-testid="foo_key"
            type="text"
            value="snack"
          />
        </span>
        <span
          class="oj-design-select"
        >
          <select
            data-testid="foo_select"
          >
            <option
              value=""
            />
            <option
              value="apple"
            >
              apple
            </option>
            <option
              value="orange"
            >
              orange
            </option>
          </select>
        </span>
        <span
          class="oj-design-expand"
        >
          <button
            data-testid="foo_expand"
            type="button"
          >
            -
          </button>
        </span>
      </div>
      <div
        class="oj-design-detail"
      >
        <div
          class="oj-design-internal-form"
        >
          <ul
            class="oj-form-group"
          >
            <li>
              <label>
                seedless
                <input
                  data-testid="foo_detail_seedless"
                  type="text"
                  value=""
                />
              </label>
            </li>
          </ul>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Test Render Some Property 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <p>
          The instance is 
          apple
        </p>
        <div
          class="oj-design-row"
        >
          <span
            class="oj-design-select"
          >
            <select
              data-testid="foo_select"
            >
              <option
                value=""
              />
              <option
                value="apple"
              >
                apple
              </option>
              <option
                value="orange"
              >
                orange
              </option>
            </select>
          </span>
          <span
            class="oj-design-expand"
          >
            <button
              data-testid="foo_expand"
              type="button"
            >
              +
            </button>
          </span>
        </div>
        <div
          class="oj-design-detail"
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <p>
        The instance is 
        apple
      </p>
      <div
        class="oj-design-row"
      >
        <span
          class="oj-design-select"
        >
          <select
            data-testid="foo_select"
          >
            <option
              value=""
            />
            <option
              value="apple"
            >
              apple
            </option>
            <option
              value="orange"
            >
              orange
            </option>
          </select>
        </span>
        <span
          class="oj-design-expand"
        >
          <button
            data-testid="foo_expand"
            type="button"
          >
            +
          </button>
        </span>
      </div>
      <div
        class="oj-design-detail"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Test Render Some Property 2`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <p>
          The instance is 
          apple
        </p>
        <div
          class="oj-design-row"
        >
          <span
            class="oj-design-select"
          >
            <select
              data-testid="foo_select"
            >
              <option
                value=""
              />
              <option
                value="apple"
              >
                apple
              </option>
              <option
                value="orange"
              >
                orange
              </option>
            </select>
          </span>
          <span
            class="oj-design-expand"
          >
            <button
              data-testid="foo_expand"
              type="button"
            >
              -
            </button>
          </span>
        </div>
        <div
          class="oj-design-detail"
        >
          <div
            class="oj-design-internal-form"
          >
            <ul
              class="oj-form-group"
            >
              <li>
                <label>
                  colour
                  <input
                    data-testid="foo_detail_colour"
                    type="text"
                    value=""
                  />
                </label>
              </li>
            </ul>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <p>
        The instance is 
        apple
      </p>
      <div
        class="oj-design-row"
      >
        <span
          class="oj-design-select"
        >
          <select
            data-testid="foo_select"
          >
            <option
              value=""
            />
            <option
              value="apple"
            >
              apple
            </option>
            <option
              value="orange"
            >
              orange
            </option>
          </select>
        </span>
        <span
          class="oj-design-expand"
        >
          <button
            data-testid="foo_expand"
            type="button"
          >
            -
          </button>
        </span>
      </div>
      <div
        class="oj-design-detail"
      >
        <div
          class="oj-design-internal-form"
        >
          <ul
            class="oj-form-group"
          >
            <li>
              <label>
                colour
                <input
                  data-testid="foo_detail_colour"
                  type="text"
                  value=""
                />
              </label>
            </li>
          </ul>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Test Render Some Property 3`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <p>
          The instance is 
          orange
        </p>
        <div
          class="oj-design-row"
        >
          <span
            class="oj-design-select"
          >
            <select
              data-testid="foo_select"
            >
              <option
                value=""
              />
              <option
                value="apple"
              >
                apple
              </option>
              <option
                value="orange"
              >
                orange
              </option>
            </select>
          </span>
          <span
            class="oj-design-expand"
          >
            <button
              data-testid="foo_expand"
              type="button"
            >
              -
            </button>
          </span>
        </div>
        <div
          class="oj-design-detail"
        >
          <div
            class="oj-design-internal-form"
          >
            <ul
              class="oj-form-group"
            >
              <li>
                <label>
                  seedless
                  <input
                    data-testid="foo_detail_seedless"
                    type="text"
                    value=""
                  />
                </label>
              </li>
            </ul>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <p>
        The instance is 
        orange
      </p>
      <div
        class="oj-design-row"
      >
        <span
          class="oj-design-select"
        >
          <select
            data-testid="foo_select"
          >
            <option
              value=""
            />
            <option
              value="apple"
            >
              apple
            </option>
            <option
              value="orange"
            >
              orange
            </option>
          </select>
        </span>
        <span
          class="oj-design-expand"
        >
          <button
            data-testid="foo_expand"
            type="button"
          >
            -
          </button>
        </span>
      </div>
      <div
        class="oj-design-detail"
      >
        <div
          class="oj-design-internal-form"
        >
          <ul
            class="oj-form-group"
          >
            <li>
              <label>
                seedless
                <input
                  data-testid="foo_detail_seedless"
                  type="text"
                  value=""
                />
              </label>
            </li>
          </ul>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
